### PR TITLE
[SPARK-23007][SQL][TEST] Add read schema suite for file-based data sources

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

The reader schema is said to be evolved (or projected) when it changed after the data is written. The followings are already supported in file-based data sources. Note that partition columns are not maintained in files. In this PR, `column` means `non-partition column`.

   1. Add a column
   2. Hide a column
   3. Change a column position 
   4. Change a column type (upcast)

This issue aims to guarantee users a backward-compatible read-schema test coverage on file-based data sources and to prevent future regressions by *adding read schema tests explicitly*.

Here, we consider safe changes without data loss. For example, data type change should be from small types to larger types like `int`-to-`long`, not vice versa.

As of today, in the master branch, file-based data sources have the following coverage.

File Format | Coverage  | Note
----------- | ---------- | ------------------------------------------------
TEXT          | N/A            | Schema consists of a single string column.
CSV            | 1, 2, 4        |
JSON          | 1, 2, 3, 4    |
ORC            | 1, 2, 3, 4    | Native vectorized ORC reader has the widest coverage among ORC formats.
PARQUET   | 1, 2, 3        |

## How was this patch tested?

Pass the Jenkins with newly added test suites.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile , @cloud-fan , @HyukjinKwon , @viirya .
Could you review this PR?

**Comment by dongjoon-hyun** [comment]
Also, ping @sameeragarwal , too.

**Comment by dongjoon-hyun** [comment]
Also, ping @rxin , too.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by gatorsmile** [comment]
We are working on the Spark 2.3 release. Could you ping us after the release?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @gatorsmile .
Is there any concern to shake 2.3 release? This is a *test case* only PR to build a clear consensus since Apache Spark 2.3.0. I think it's safe to be part of 2.3.0.

**Comment by gatorsmile** [comment]
Do not have enough review bandwidth on this test-only PRs before Spark 2.3 release

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: nit: `` `byte` `` -> `byte` or the opposite for consistency with the same instances.

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: Here seems many tests have some duplicated codes .. can we maybe do such as something like as below?

```scala
Seq(byteDF, ...).zip("byte").foreach { case (df, t) =>
  test(s"boolean to $t") {
    spark.read
      .schema("col1 long")
      .format(format)
      .options(options)
      .load(path)
    checkAnswer(df4, longDF)
  }
}
```

I am fine with any idea to deal with this duplication.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: I think we can do `withTempPath`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: Shall we leave the number given above in this comment like `(case 1.)`.

**Comment by HyukjinKwon** [comment]
cc @sameeragarwal for reviewing too. I vaguely remember we had a talk about this before.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @HyukjinKwon . I'll update like that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: Ur, for this, when we put the variables (byteDF, ...) outside of `test` functions, it seems to cause SQLContext errors.

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile , @cloud-fan , @sameeragarwal , @HyukjinKwon .
The PR is ready for review again. Spark commit log seems to be a little quiet since yesterday.
Could you squeeze some time to give for this Schema Evolution suite? Thank you in advance for any advice!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: @dongjoon-hyun, how do we guarantee schema change in Parquet and ORC?

I thought we (roughly) randomly pick up a file, read its footer and then use it. So, I was thinking we don't properly support this. It makes sense to Parquet with `mergeSchema` tho.

I think it's not even guaranteed in CSV too because we will rely on its header from one file.

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: Ohaaa, the schema is explicitly set here. Sorry, I missed it.

**Comment by gatorsmile** [comment]
Will do it after 2.3 release

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin , @cloud-fan , @sameeragarwal , @HyukjinKwon .

Could you give me some opinions about this PR? I know that Xiao Li is busy for this period, so I didn't ping hime. For me, this PR is important. Sorry for annoying you guys.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Rebased to the master.

**Comment by dongjoon-hyun** [comment]
Finally, Spark 2.3 passes the vote. Could you review this, @gatorsmile ?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for retriggering, @HyukjinKwon !

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile , @HyukjinKwon , @cloud-fan .
Since 2.3 is officially announced, I'm pinging you guys again. :)
Please let me know if there is something for me to do here.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by HyukjinKwon — APPROVED** [review]
LGTM but let me leave it to @gatorsmile and @cloud-fan.

**Comment by dongjoon-hyun** [comment]
Thank you so much for review and approval, @HyukjinKwon !

**Comment by gatorsmile** [comment]
Before adding the test cases, schema evolution is officially supported? Could you describe it in details?

**Comment by dongjoon-hyun** [comment]
@gatorsmile . Yes. It's officially supported Hive like the following compatible cases which means upgrading wider types. As you see in my PR, Apache Spark 2.3 also supports, but the degree of support depends on data sources.

```sql
hive> CREATE TABLE t1(a int) STORED AS ORC;

hive> INSERT INTO t1 VALUES(1);

hive> ALTER TABLE t1 CHANGE a a bigint;

hive> SELECT * FROM t1;
OK
1
Time taken: 0.137 seconds, Fetched: 1 row(s)
```

**Comment by dongjoon-hyun** [comment]
After merging this PR, I'm going to make a PR on SQL part for Apache Spark 2.4.

**Comment by gatorsmile** [comment]
Basically, these commands are not issued from Spark. We only have the support for ALTER TABLE ADD COLUMN. This PR is just testing the capability of our automatic schema inference, right?

**Comment by dongjoon-hyun** [comment]
Of course! Those command are an illustration of use case of schema evolution in Hive-side in general.

This PR aims to provide *schema evolution* test coverage in Spark-side. As you see, this is for only `sql/core`, not for `sql/hive` module at all.

Apache Spark 2.3 already supports the same schema evolution in terms of **userGivenSchema**. (In Hive case, it's HMS schema instead of file schema.)
```scala
val df = spark.read
        .schema(userGivenSchema)
        .format(format)
        .options(options)
        .load(path)
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala]: @gatorsmile . Please see this. This is not about **schema inferencing**.

**Comment by gatorsmile** [comment]
Let us write the document in this PR? Document our current behaviors with the related test cases.

**Comment by dongjoon-hyun** [comment]
Which document do you have in mind? A section in `docs/sql-programming-guide.md`?

**Comment by gatorsmile** [comment]
Yes

File: docs/sql-programming-guide.md
Comment (dongjoon-hyun) [inline docs/sql-programming-guide.md]: @gatorsmile . I rebased to the master and added this.

File: docs/sql-programming-guide.md
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: In SQL standard, when we remove a column, all the data are removed. However, we do not support it. Users could still see the data after they add the column with the same name like what they removed previously.

File: docs/sql-programming-guide.md
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: Do we support it? When people issuing `select * from tab`, we automatically reorder the partition columns to the end of the schema.

File: docs/sql-programming-guide.md
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: These are just upcast.

File: docs/sql-programming-guide.md
Comment (gatorsmile) [reply docs/sql-programming-guide.md]: Based on the current behavior, we do not support schema evolution. Schema evolution is a well-defined term. It sounds like this PR is try to test the behaviors when users provide the schema that does not exactly match the physical schema. This is different from the definition of schema evolution.

Comment (dongjoon-hyun) [reply docs/sql-programming-guide.md]: Thank you so much for review, @gatorsmile . I waited for this moment. :)
I agree all of your comments. The main reason of those limitation is because Spark file-based data sources doesn't have a capability to manage multi-version schema and the column default values here. In fact, that is beyond of Spark data sources' role. Thus, this PR is trying to add a test coverage for AS-IS capability in order to prevent future regression and to make a foundation to trust and to build on later. I don't think this is worthy of documentation at the beginning. It's a start.

File: docs/sql-programming-guide.md
Comment (dongjoon-hyun) [reply docs/sql-programming-guide.md]: Right. The test case doen't aim to cover those cases so far.

File: docs/sql-programming-guide.md
Comment (dongjoon-hyun) [reply docs/sql-programming-guide.md]: Correct, we need to clarify that partition columns are always at the end.

File: docs/sql-programming-guide.md
Comment (dongjoon-hyun) [reply docs/sql-programming-guide.md]: Yep. `Upcast`s are safe. This PR doesn't aim to cover or guarantee unsafe casting at this stage. Although these are straight-forward `upcast`s, not all Spark file-based data sources seems to support them (based on the test cases). This PR is trying to set the clear boundary and to clarify those missed things.

**Comment by dongjoon-hyun** [comment]
I'll update like the followings.
- Remove `Remove a column` part from the description parts (docs/testsuite file doc) while keeping the test cases.
- Add a clear description about `partition columns` position rules.
- Mention `upcast` for `Change a column type` part.

For `docs/sql-programming-guide.md`, I'll keep during review period.

**Comment by dongjoon-hyun** [comment]
Rebased to the master.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: docs/sql-programming-guide.md
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: I still want to avoid using `schema evolution` in the doc or tests. `Schema Projection` might better. More importantly, you have to clarify that this only covers the read path.

What is the behavior in the write path when the physical and data schemas are different.

Comment (dongjoon-hyun) [reply docs/sql-programming-guide.md]: Thank you for review, @gatorsmile . I'll update like that.

For write operation, we cannot specify schema like read path, `.schema`. Spark writes the new file into the directory additionally or overwrites the directory.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaSuite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaSuite.scala]: @gatorsmile . Now, it becomes `ReadSchemaSuite`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala]: Here, I used `evolved` and `projected` as a general verb.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala]: Thanks for helping improve the test coverage! All the included test cases are positive. How about the negative test cases? What kind of errors you hit?

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala]: Yes. Right. Since the main purpose of this PR is preventing regressions, it consists of positive-only. The errors are case-by-case for each data sources.

For `BooleanTypeTest` example, Parquet raises higher exceptions due to `ClassCastException` (at the bottom). JSON raises `Results do not match` test case failures without exceptions.

- Parquet
```
org.apache.spark.sql.execution.QueryExecutionException: Encounter error while reading parquet files. One possible cause: Parquet column cannot be converted in the corresponding files. Details: 
...
Caused by: org.apache.parquet.io.ParquetDecodingException: Can not read value at 1 in block 0 in file file:/private/var/folders/dc/1pz9m69x14q_gw8t7m143t1c0000gn/T/spark-4b3d788b-1d7e-4ca2-9c01-88f639daf02f/part-00000-975391e5-1f1d-49f5-8e12-3213281618ed-c000.snappy.parquet
...
Caused by: java.lang.ClassCastException: org.apache.spark.sql.catalyst.expressions.MutableByte cannot be cast to org.apache.spark.sql.catalyst.expressions.MutableBoolean
```

**Comment by dongjoon-hyun** [comment]
The test suite is designed like the following according to their features.
```scala
class CSVReadSchemaSuite
  extends ReadSchemaSuite
  with IntegralTypeTest
  with ToDoubleTypeTest
  with ToDecimalTypeTest
  with ToStringTypeTest {

  override val format: String = "csv"
}
```
To add a negative test case, we need to do something like `with NoBooleanTypeTest`. How do you think about that?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
@gatorsmile . Please let me know if I need to do more.

**Comment by gatorsmile** [comment]
@dongjoon-hyun This PR is to improve the test coverage. LGTM. 

When the schema do not match with the schemas of underlying data source, the current error messages might be weird. This is a common issue, I think. Please submit a separate PR to improve the error handling in these cases?

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @gatorsmile . Sure. I'll make a PR to improve error handling for that.

**Comment by dongjoon-hyun** [comment]
Also, thank you, @HyukjinKwon .

**Pull request closed without merging**
